### ISSUE TITLE
Add CustomAnisotropicNonbondedForce 

### ISSUE THREAD

**Issue by mvanvleet** [pandegroup/openmm#1545]
I'm looking to implement a pairwise non-bonded force that depends on both the inter-particle distance and on angles that can be defined with respect to a local reference frame for each atom. Github doesn't do well with math equations, so I've attached a pdf that explains the functional form in detail:

[pairwise_int_potential.pdf](https://github.com/pandegroup/openmm/files/377062/pairwise_int_potential.pdf)

For reasons listed in the document, none of the existing CustomForce functions are sufficient for implementing this functional form. I intend to add a CustomAnisotropicNonbondedForce class which would extend CustomNonbondForce to allow for the following:

-  In addition to the inter-particle distance r, angles theta12, theta21, phi12, and phi21 would be accepted as arguments in the energy expression.
-  Identically to AmoebaMultipoleForce, local axis definitions could be read in for each particle; these local axis definitions would define the angles listed above.

I'm planning on implementing this enhancement feature myself (starting from CustomNonbondedForce), however am posting this here for two reasons:

1. This will be my first time developing in OpenMM, so suggestions from more experienced developers are always welcome.
2. I know a few others in the community that are experimenting with (or are interested in using) anisotropic potentials, and am willing to entertain additional features to the class if anyone is working with similar potentials.

**Comment by rmcgibbo** [comment]
Out of curiosity, what are your plans for cutoffs / long range behavior? It's not so important for anisotropic exchange, but a `CustomAnisotropicNonbondedForce` wouldn't be limited to exponentially decaying functions in `r_12`.

**Comment by mvanvleet** [comment]
For implementing long-range behavior in `CustomAnisotropicNonbondedForce`, I'll probably follow the behavior of `CustomNonbondedForce` and only allow for periodic or non-periodic cutoffs.

For my force fields, however (and for general simulations), I recognize that I'll need a long-range correction of some sort to treat dispersion or other long-range effects. Fortunately for my systems, the anisotropic correction (which would be difficult to calculate!) is probably well-approximated by the isotropic long-range correction. Depending on if/when Lennard-Jones PME gets implemented, I'll either use a PME approach or a standard long-range correction factor to treat long-range effects.

So in summary, I wasn't planning to include a sophisticated treatment of long-range effects in `CustomAnisotropicNonbondedForce`, mostly because I have no good general solution for long-range corrections. If you have suggestions, though, I'd be happy to hear them!

**Comment by peastman** [comment]
You mention that neither CustomManyParticleForce or CustomHbondForce is flexible enough to meet your needs.  It's not obvious to me why.  What features are they missing that would be needed for this?

**Comment by mvanvleet** [comment]
Peter,

My force fields require the computation of 4 different angles: theta12, theta21, phi12, and phi21. These are most easily computed by constructing a local reference frame for each atom, however one could also place virtual sites on each atom to represent the z- and x-axis, respectively (which is what one would need to do to use CustomManyParticleForce or CustomHbondForce). In this case, there are 6 sites that need to be considered in each pairwise interaction.

To use CustomManyParticleForce, the energy expression either needs to be symmetric with respect to all particle permutations, or symmetric to permutations allowing for one central (i.e. unique) atom. My force field depends on two 'central/unique' atoms, and the expressions are certainly not symmetric with respect to the interchange of particle axes. For this reason, CustomManyParticleForce does not seem to be a good choice.

CustomHbondForce is better, and allows for the computation of angles like I need, however it will only (as far as I can tell) compute donor-acceptor interactions. Even if I arbitrarily define certain atoms as 'donors' and others as 'acceptors', I would need to compute donor-donor and acceptor-acceptor interactions, which CustomHbondForce does not do. I expect that this could be accomplished with fairly minimal modifications, and I will likely look to CustomHbondForce as a second example on which to base CustomAnisotropicNonbondedForce.
